api/cryptocom/client.go

```diff
@@ -0,0 +1,271 @@
+        if err != nil {
+          continue
+        }
+
+        time.Sleep(3 * time.Second) // Cryptocom requires this sleep
+        newCnx := Connection{Endpoint: cnx.Endpoint, IsPrivate: cnx.IsPrivate, Transport: conn}
+
+        if newCnx.IsPrivate {
+          c.privateConn = newCnx
+          c.authenticate()
+          if len(c.privateSubs) > 0 {
+            c.subscribePrivateChannels(c.privateSubs)
+          }
+        } else {
+          c.publicConn = newCnx
+          if len(c.publicSubs) > 0 {
+            c.subscribePublicChannels(c.publicSubs)
+          }
+        }
+
+        cnx.Close()
+        cnx = newCnx
+        break
+      }
+
+      continue
+    }
+
+    c.LogFunc("Received [%s]: %s\n", cnx.Type(), string(m))
+
+    var parsed Response
+    err = json.Unmarshal(m, &parsed)
+    if err != nil {
+      c.LogFunc("error on JSON.Unmarshal")
+      continue
+    }
+
+    if parsed.Method == "public/heartbeat" {
+      c.respondHeartBeat(cnx.IsPrivate, parsed.Id)
+      continue
+    }
+
+    c.outbox <- parsed
+  }
+}
+
+func (c *Client) generateSignature(r *Request) {
+  secret := c.secret
+  concatenedParams := ""
+
+  var parameters []string
+  for key := range r.Params {
+    parameters = append(parameters, key)
+  }
+
+  sort.Strings(parameters)
+
+  for _, v := range parameters {
+    concatenedParams += v + r.Params[v].(string)
+  }
+
+  data := r.Method + strconv.Itoa(r.Id) + r.ApiKey + concatenedParams + r.Nonce
+  h := hmac.New(sha256.New, []byte(secret))
+  h.Write([]byte(data))
+  sha := hex.EncodeToString(h.Sum(nil))
+  r.Signature = sha
+}
+
+func (c *Client) authenticate() {
+  r := c.AuthRequest()
+  c.sendPrivateRequest(r)
+}
+
+func (c *Client) subscribePrivateChannels(channels []string) error {
+  r := c.subscribeRequest(channels)
+  return c.sendPrivateRequest(r)
+}
+
+func (c *Client) subscribePublicChannels(channels []string) error {
+  r := c.subscribeRequest(channels)
+  return c.sendPublicRequest(r)
+}
+
+func (c *Client) sendPrivateRequest(r *Request) error {
+  defer c.privateConn.Unlock()
+  b, err := r.Encode()
+
+  if err != nil {
+    return err
+  }
+
+  c.LogFunc("Sending private: %s\n", string(b))
+
+  c.privateConn.Lock()
+  return c.privateConn.WriteMessage(websocket.TextMessage, b)
+}
+
+func (c *Client) sendPublicRequest(r *Request) error {
+  defer c.publicConn.Unlock()
+  b, err := r.Encode()
+
+  if err != nil {
+    return err
+  }
+  c.LogFunc("Sending public: %s\n", string(b))
+
+  c.publicConn.Lock()
+  return c.publicConn.WriteMessage(websocket.TextMessage, b)
+}
+
+func (c *Client) recordPublicSubscription(channels []string) {
+
+  c.publicSubs = append(c.publicSubs, channels...)
+
+}
+
+func (c *Client) recordPrivateSubscription(channels []string) {
+
+  c.privateSubs = append(c.privateSubs, channels...)
+}
+
+func isClosedCnxError(err error) bool {
+  return strings.Contains(err.Error(), "use of closed network connection")
```

Comment:
https://pkg.go.dev/github.com/gorilla/websocket#IsCloseError

Reply:
We cannot use it here.
IsCloseError is not a method of websocket client. It will look strange if we use interface variable everywhere, but then we use websocket.